src/lib.rs

```diff
@@ -308,3 +308,117 @@ pub fn narrowable_abgc(args: TokenStream, input: TokenStream) -> TokenStream {
 
     TokenStream::from(expanded)
 }
+
+#[cfg(feature = "gcmalloc")]
+#[proc_macro_attribute]
+pub fn narrowable_gcmalloc(args: TokenStream, input: TokenStream) -> TokenStream {
+    let args = parse_macro_input!(args as AttributeArgs);
+    let input = parse_macro_input!(input as ItemTrait);
+    if args.len() != 1 {
+        panic!("Need precisely one argument to 'narrowable'");
+    }
+    let struct_id = match &args[0] {
+        NestedMeta::Meta(m) => m.name(),
+        NestedMeta::Literal(_) => panic!("Literals not valid attributes to 'narrowable'")
+    };
+    let trait_id = &input.ident;
+    let expanded = quote! {
+        /// A narrow pointer to #trait_id.
+        pub struct #struct_id {
+            // This struct points to a vtable pointer followed by an object. In other words, on a
+            // 64 bit machine the layout is (in bytes):
+            //   0..7: vtable
+            //   8..: object
+            // This is an inflexible layout, since we can only support structs whose alignment is
+            // the same or less than a usize's.
+            // FIXME: This is somewhat a hangover from abgc. Why can't it
+            // support interior pointers, and does gcmalloc need the same
+            // restrictions?
+            vtable: *mut u8
+        }
+
+        impl #struct_id {
+            /// Create a new narrow pointer to #trait_id.
+            pub fn new<U>(v: U) -> ::gcmalloc::Gc<Self>
+            where
+                *const U: ::std::ops::CoerceUnsized<*const (dyn #trait_id + 'static)>,
+                U: #trait_id + 'static
+            {
+                let (layout, uoff) = ::std::alloc::Layout::new::<usize>().extend(
+                    ::std::alloc::Layout::new::<U>()).unwrap();
+                // Check that we've not been given an object whose alignment exceeds that of a
+                // size: we can't handle such cases until abgc can store interior pointers.
```

Comment:
This is the place to reconsider the interior pointer stuff I suspect.

Reply:
Perhaps we should merge this and I can revisit interior pointers properly with some tests and examples in `gcmalloc`. My current understanding is that interior pointers in gcmalloc should "just work". I've certainly written code to deal with them, but admittedly they are less well tested than other areas.

Reply:
Agreed.